clients/src/main/java/org/apache/kafka/common/telemetry/internals/KafkaMetricsCollector.java

```diff
@@ -0,0 +1,347 @@
+            } else {
+                // skip non-measurable metrics
+                log.debug("Skipping non-measurable gauge metric {}", metricKey.getName());
+            }
+        }
+    }
+
+    private void collectDelta(MetricKey originalKey, Double value, MetricsEmitter metricsEmitter) {
+        MetricKey metricKey = metricNamingStrategy.derivedMetricKey(originalKey, "delta");
+        if (!metricsEmitter.shouldEmitMetric(metricKey)) {
+            return;
+        }
+
+        // calculate a getAndSet, and add to out if non-empty
+        final Instant timestamp = clock.instant();
+        InstantAndValue<Double> instantAndValue = ledger.delta(originalKey, timestamp, value);
+
+        metricsEmitter.emitMetric(
+            SinglePointMetric.deltaSum(metricKey, instantAndValue.getValue(), true, timestamp,
+                instantAndValue.getIntervalStart())
+        );
+    }
+
+    private void collectSum(MetricKey metricKey, double value, MetricsEmitter metricsEmitter) {
+        if (!metricsEmitter.shouldEmitMetric(metricKey)) {
+            return;
+        }
+
+        metricsEmitter.emitMetric(
+            SinglePointMetric.sum(metricKey, value, true, clock.instant(), ledger.instantAdded(metricKey))
+        );
+    }
+
+    private void collectGauge(MetricKey metricKey, Number value, MetricsEmitter metricsEmitter) {
+        if (!metricsEmitter.shouldEmitMetric(metricKey)) {
+            return;
+        }
+
+        metricsEmitter.emitMetric(
+            SinglePointMetric.gauge(metricKey, value, clock.instant())
+        );
+    }
+
+    @Override
+    public String toString() {
+        return this.getClass().getCanonicalName();
+    }
+
+    private static boolean isMeasurable(KafkaMetric metric) {
+        // KafkaMetric does not expose the internal MetricValueProvider and throws an IllegalStateException exception
+        // if .measurable() is called for a Gauge.
+        // There are 2 ways to find the type of internal MetricValueProvider for a KafkaMetric - use reflection or
+        // get the information based on whether or not a IllegalStateException exception is thrown.
+        // We use reflection so that we can avoid the cost of generating the stack trace when it's
+        // not a measurable.
+        try {
+            Object provider = METRIC_VALUE_PROVIDER_FIELD.get(metric);
+            return provider instanceof Measurable;
+        } catch (Exception e) {
+            throw new KafkaException(e);
+        }
+    }
+
+    /**
+     * Keeps track of the state of metrics, e.g. when they were added, what their getAndSet value is,
+     * and clearing them out when they're removed.
+     */
+    private class StateLedger {
+
+        private final Map<MetricKey, KafkaMetric> metricMap = new ConcurrentHashMap<>();
+        private final LastValueTracker<Double> doubleDeltas = new LastValueTracker<>();
+        private final ConcurrentMap<MetricKey, Instant> metricAdded = new ConcurrentHashMap<>();
```

Comment:
Why do we need a `ConcurrentMap`, and why do we clear the type not just as `Map` (similar to `metricMap`)?

Reply:
metrics can be removed in a different thread than the one collecting metrics

Reply:
I have moved them to Map from ConcurrentMap, though implementation is ConcurrenHashMap.